man/cryptsetup.8
.TH CRYPTSETUP "8" "March 2005" "cryptsetup 1.0.3" "Maintainance Commands"
.SH NAME
cryptsetup - setup cryptographic volumes for dm-crypt (including LUKS extension)
.SH SYNOPSIS

.B cryptsetup <options> <action> <action args>

.SH DESCRIPTION
.\" Add any additional description here
.PP
cryptsetup is used to conveniently setup dm-crypt managed device-mapper mappings. For basic dm-crypt mappings, there are five operations.
.SH ACTIONS
These strings are valid for \fB<action>\fR, followed by their \fB<action args>\fR:

\fIcreate\fR <name> <device>
.IP
creates a mapping with <name> backed by device <device>.
<options> can be [\-\-hash, \-\-cipher, \-\-verify-passphrase, \-\-key-file, \-\-key-size, \-\-offset, \-\-skip, \-\-readonly]
.PP
\fIremove\fR <name>
.IP
removes an existing mapping <name>. No options.
.PP
\fIstatus\fR <name>
.IP
reports the status for the mapping <name>. No options.
.PP
\fIreload\fR <name>
.IP
modifies an active mapping <name>. Same options as for create.
.PP
\fIresize\fR <name>
.IP
resizes an active mapping <name>. <options> must include \-\-size 
.PP
.br
.SH LUKS EXTENSION

LUKS, Linux Unified Key Setup, is a standard for hard disk encryption. It standardizes a partition header, as well as the format of the bulk data. LUKS can manage multiple passwords, that can be revoked effectively and that are protected against dictionary attacks with PBKDF2. 

These are valid LUKS actions:

\fIluksFormat\fR <device> [<key file>]
.IP
initializes a LUKS partition and sets the initial key, either via prompting or via <key file>.
<options> can be [\-\-cipher, \-\-verify-passphrase, \-\-key-size, \-\-key-slot].
.PP
\fIluksOpen\fR <device> <name>
.IP
opens the LUKS partition <device> and sets up a mapping <name> after successful verification of the supplied key material (either via key file by \-\-key-file, or via prompting).
<options> can be [\-\-key-file, \-\-readonly].
.PP
\fIluksClose\fR <name>
.IP
identical to \fIremove\fR.
.PP
\fIluksAddKey\fR <device> [<new key file>]
.IP
add a new key file/passphrase. An existing passphrase or key file (via \-\-key-file) must be supplied. The key file with the new material is supplied as a positional argument. <options> can be [\-\-key-file, \-\-key-slot].
.PP
\fIluksKillSlot\fR <device> <key slot number>
.IP
wipe key with number <key slot> from LUKS device. A remaining passphrase or key file (via \-\-key-file) must be supplied. <options> can be [\-\-key-file].
.PP
\fIluksDelKey\fR <device> <key slot number>
.IP
identical to luksKillSlot, but deprecated action name.
.PP
\fIluksUUID\fR <device>
.IP
print UUID, if <device> has a LUKS header. No options.
.PP
\fIisLuks\fR <device>
.IP
returns true, if <device> is a LUKS partition. Otherwise, false. No options.
.PP
\fIluksDump\fR <device>
.IP
dumps the header information of a LUKS partition. No options.
.PP

For more information about LUKS, see \fBhttp://luks.endorphin.org\fR
.SH OPTIONS
.TP
.B "\-\-hash, \-h"
specifies hash to use for password hashing. This option is only relevant for the "create" action. The hash string is passed to libgcrypt, so all hashes accepted by gcrypt are supported. Default is "ripemd160".
.TP
.B "\-\-cipher, \-c"
set cipher specification string. Usually, this is "aes-cbc-plain". For pre-2.6.10 kernels, use "aes-plain" as they don't understand the new cipher spec strings. To use ESSIV, use "aes-cbc-essiv:sha256".
.TP
.B "\-\-verify-passphrase, \-y"
query for passwords twice. Useful when creating a (regular) mapping for the first time, or when running \fIluksFormat\fR.
.TP
.B "\-\-key-file, \-d"
use file as key material. With LUKS, key material supplied in key files via \-d are always used for existing passphrases. If you want to set a new key via a key file, you have to use a positional arg to \fIluksFormat\fR or \fIluksAddKey\fR.

If the key file is "-", stdin will be used. This is different from how cryptsetup usually reads from stdin. See section \fBNOTES ON PASSWORD PROCESSING\fR for more information.
.TP
.B "\-\-key-slot, \-S"
For LUKS operations that add key material, this options allows to you specify which key slot is selected for the new key. This option can be used for luksFormat and luksAddKey.
.TP
.B "\-\-key-size, \-s"
set key size in bits. Has to be a multiple of 8 bits. The key size is limited by the used cipher. See output of /proc/crypto for more information. Can be used for \fIcreate\fR or \fIluksFormat\fR, all other LUKS actions will ignore this flag, as the key-size is specified by the partition header. Default is 128.
.TP
.B "\-\-size, \-b"
force the size of the underlying device in sectors.
.TP
.B "\-\-offset, \-o"
start offset in the backend device.
.TP
.B "\-\-skip, \-p"
how many sectors of the encrypted data to skip at the beginning. This is different from the \-\-offset options with respect to IV calculations. Using \-\-offset will shift the IV calculation by the same negative amount. Hence, if \-\-offset \fIn\fR, sector \fIn\fR will be the first sector on the mapping with IV \fI0\fR. Using \-\-skip would have resulted in sector \fIn\fR being the first sector also, but with IV \fIn\fR.
.TP
.B "\-\-readonly"
set up a read-only mapping.
.TP
.B "\-\-iter-time, \-i"
The number of milliseconds to spend with PBKDF2 password processing. This option is only relevant to the LUKS operations as \fIluksFormat\fR or \fIluksAddKey\fR.
.TP
.B "\-\-batch-mode, \-q"
Do not ask for confirmation. This option is only relevant for \fIluksFormat\fR.
.TP
.B "\-\-timeout, \-t"
The number of seconds to wait before timeout. This option is relevant every time a password is asked, like \fIcreate\fR, \fIluksOpen\fR, \fIluksFormat\fR or \fIluksAddKey\fR. It has no effect if used in conjunction with \-\-key-file.
.TP
.B "\-\-tries, \-T"
How often the input of the passphrase shall be retried. This option is relevant every time a password is asked, like \fIcreate\fR, \fIluksOpen\fR, \fIluksFormat\fR or \fIluksAddKey\fR. The default is 3 tries.
.TP
.B "\-\-align-payload=\fIvalue\fR"
Align payload at a boundary of \fIvalue\fR 512-byte sectors. This option is relevant for \fIluksFormat\fR.  If your block device lives on a RAID, it is
useful to align the filesystem at full stripe boundaries so it can take advantage of the RAID's geometry.  See for instance the sunit and swidth options
in the mkfs.xfs manual page. By default, the payload is aligned at an 8 sector (4096 byte) boundary.
.TP
.B "\-\-version"
Show the version.

.SH NOTES ON PASSWORD PROCESSING
\fIFrom a file descriptor or a terminal\fR: Password processing is new-line sensitive, meaning the reading will stop after encountering \\n. It will process the read material (without newline) with the default hash or the hash given by \-\-hash. After hashing, it will be cropped to the key size given by \-s (default 256 bits).

\fIFrom stdin\fR: Reading will continue until EOF (so using e.g. /dev/random as stdin will not work), with the trailing newline stripped. After that the read data will be hashed with the default hash or the hash given by \-\-hash and the result will be cropped to the keysize given by \-s (default 256 bits). If "plain" is used as an argument to the hash option, the input data will not be hashed. Instead it will be zero padded (if shorter than the keysize) or truncated (if longer than the keysize) and used directly as the key.
Instead, it will be zero padded (if shorter than the keysize) or truncated (if longer than the keysize) and used directly as the key. No warning will be given if the amount of data read from stdin is less than the keysize.

\fIFrom a key file\fR: It will be cropped to the size given by \-s. If there is insufficient key material in the key file, cryptsetup will quit with an error.

If \-\-key-file=- is used for reading the key from stdin, no trailing newline is stripped from the input. Without that option, cryptsetup strips trailing newlines from stdin input.
.SH NOTES ON PASSWORD PROCESSING FOR LUKS
LUKS uses PBKDF2 to protect against dictionary attacks (see RFC 2898). 
LUKS will always use SHA1 in HMAC mode, and no other mode is supported at the moment. 
Hence, \-h is ignored.

LUKS will always do an exhaustive password reading. Hence, password can not be read from /dev/random, /dev/zero or any other stream that does not terminate.

LUKS saves the processing options when a password is set to the respective key slot.
Therefore, no options can be given to luksOpen. 
For any password creation action (luksAddKey, or luksFormat), the user may specify how much the time the password processing should consume.
Increasing the time will lead to a more secure password, but also will take luksOpen longer to complete. The default setting of one second is sufficient for good security.
.SH NOTES ON PASSWORDS
Mathematic can't be bribed. Make sure you keep your passwords safe. There are a few nice tricks for constructing a fallback, when suddenly out of (or after being) blue, your brain refuses to cooperate. These fallbacks are possible with LUKS, as it's only possible with LUKS to have multiple passwords.
.SH AUTHORS
cryptsetup is written by Christophe Saout <christophe@saout.de>
.br
LUKS extensions, and man page by Clemens Fruhwirth <clemens@endorphin.org>
.SH "COMPATABILITY WITH OLD SUSE TWOFISH PARTITIONS"
To read images created with SuSE Linux 9.2's loop_fish2 use \-\-cipher
twofish-cbc-null \-s 256 \-h sha512, for images created with even
older SuSE Linux use \-\-cipher twofish-cbc-null \-s 192 \-h
ripemd160:20
.SH "REPORTING BUGS"
Report bugs to <dm-crypt@saout.de>.
.SH COPYRIGHT
Copyright \(co 2004 Christophe Saout
.br
Copyright \(co 2004-2006 Clemens Fruhwirth

This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH "SEE ALSO"

dm-crypt website, \fBhttp://www.saout.de/misc/dm-crypt/\fR

LUKS website, \fBhttp://luks.endorphin.org\fR

dm-crypt TWiki, \fBhttp://www.saout.de/tikiwiki/tiki-index.php\fR


debian/changelog
cryptsetup (2:1.0.6-1ubuntu3) intrepid; urgency=low

  * Parse comments in lines not starting with '#', LP: #185380
  * in cryptroot hook, don't rely on 'udevadm settle' to wait long enough
    for the cryptdevice to appear. Reimplement the busy waiting loop found
    while waiting for the root file system. Patch based on work by Swâmi
    Petaramesh. LP: #164044
  * debian/crypdisks.functions: call 'env' with full path. LP: #178829.

 -- Reinhard Tartler <siretart@tauware.de>  Mon, 26 May 2008 22:12:32 +0200

cryptsetup (2:1.0.6-1ubuntu2) intrepid; urgency=low

  * Simplify the patch in debian/cryptdisks.functions that stops usplash
    before asking for a passphrase.

 -- Reinhard Tartler <siretart@tauware.de>  Mon, 26 May 2008 20:18:14 +0200

cryptsetup (2:1.0.6-1ubuntu1) intrepid; urgency=low

  * Merge new debian version. Remaining changes:
    - cryptsetup is linked dynamically against libgcrypt and libgpg-error.
    - stop usplash on user input. LP #62751
    - debian/cryptdisks.functions: Always output and read from the console.
      LP #58794.
    - Add XSBC-Vcs-Bzr tag to indicate that this package is managed using
      bzr on launchpad.
    - debian/initramfs/cryptroot-hook: LP #73862
      Added patch to install aes optimized cypher module
    - try to load optimized cypher module in cryptsetup.functions as well,
      because cryptroot-hook is only executed when we really have a
      cryptoroot.
  * other ubuntu changes have been merged into debian. Please report bugs
    if you believe some patches have been dropped.
  * removed 07_typos_fix.patch, has been reviewed and applied upstream.

 -- Reinhard Tartler <siretart@tauware.de>  Sun, 25 May 2008 22:52:30 +0200

cryptsetup (2:1.0.6-1) unstable; urgency=low

  [ Jonas Meurer ]
  * new upstream release
    - reload option is depreciated and a warning is printed. (closes: #428288)
  * convert patch system from dpatch to quilt.
  * enhance the information regarding the default hash setting in NEWS.Debian.
    Thanks to Ross Boylan <ross@biostat.ucsf.edu>.
  * change author of keyslot patch to Marc Merlin in changelog, thanks to
    U. Kuehn for raising that issue.
  * doing some debian/rules redesign and cleanup, speeds up the build process.
  * ignore devices with the noauto option early enough to prevent any checks
    on them. Thanks to Joachim Breitner <nomeata@debian.org> (closes: #464672)
  * update debian/copyright to actually mention copyright, thanks lintian.
  * change script=$(basename $req) to script=${req##*/} in initramfs cryptroot
    script. Thanks to Adeodato Simó <dato@net.com.org.es>. (closes: #466240)
  * change test ... -a ... to [ ... ] && [ ... ] in the check scripts.
  * add support for tries option to initramfs scripts. Thanks to Helmut Grohne
    <helmut@subdivi.de>. (closes: #430158, #469869) Use --tries=1 for
    cryptsetup in the initramfs script. Document the difference between
    initscript and initramfs for tries=0 in the crypttab manpage.
  * add, build and install askpass.c, a helper program by David Härdeman. The
    idea is to use it for passphrase prompt in the initramfs script.

  [ David Härdeman ]
  * Work with LABEL=, UUID= and symlinks in /etc/fstab (closes: #466175)
  * Improve module loading in initramfs hook so that the newer  as well
    as arch specific crypto drivers are taken into consideration
    (closes: #464673)
  * Depend on race-free version of libdevmapper, thus making udevsettle
    call from cryptsetup binary unnecessary. Also change call to
    udevsettle in initramfs script (which is still useful as it related
    to the source device) to optionally use udevadm if present (closes:
    #456326).

 -- Jonas Meurer <mejo@debian.org>  Mon, 31 Mar 2008 15:58:35 +0200

cryptsetup (2:1.0.6~pre1+svn45-1) unstable; urgency=low

  * New upstream svn snapshot:
    - Adds typo fixes by Justin Pryzby <jpryzby+d@quoininc.com> to cryptsetup.8
      manpage.
    - Mentions luksKillSlot in the manpage. Thanks to Alexander Heinlein
      <alexander.heinlein@web.de>. (closes: #459206)
    - Adds the patch by Marc Merlin <marc_www@merlins.org> to support explicit
      key slots for luksFormat and luksAddKey. Thanks to U. Kuehn, who figured
      out that this patch wasn't applied even though changelog said so.
    - Supports adding new keys to active devices again. Thanks to Tobias Frost
      <tobi@coldtobi.de> for the bugreport. (closes: #460409)
  * Add support for a custom filesystem for /tmp. Patch provided by
    Hans-Peter Oeri <hp@oeri.ch>.
  * Add X-Start-Before headers to cryptdisks and cryptdisks-early initscripts.
    Thanks to Petter Reinholdtsen <pere@debian.org> for report and patch.
    (closes: #458944)
  * Add support for a noauto option to cryptdisks. Thanks to U Kuehn
    <ukuehn@acm.org> for the idea.
  * Add typo fixes by Justin Pryzby <jpryzby+d@quoininc.com> to crypttab.5
    manpage. (closes: #460994)
  * Add a cryptdisks_stop script, corresponding to cryptdisks_start. Thanks to
    Joachim Breitner <nomeata@debian.org> for the idea. (closes: #459832)
  * Change log_progress_msg to log_action_msg in cryptdisks.functions. That
    way a newline is printed after the start of every device. Thanks to Frans
    Pop <elendil@planet.nl> for the bugreport. (closes: #461548)
  * Add bash_completition script provided by Kevin Locke <kwl7@cornell.edu>.
    (closes: #423591)
  * Fix a spelling error in the package description: linux -> Linux.
  * Fix bashisms in cryptdisks_{start,stop} found by Raphael Geissert
    <atomo64+debian@gmail.com>.
  * Change the default hash in initramfs scripts from sha256 to ripemd160 for
    consistency with cryptsetup default. Add information about that to
    NEWS.Debian. Thanks to martin f krafft <madduck@debian.org>.
    (closes: #406317)

 -- Jonas Meurer <mejo@debian.org>  Wed, 30 Jan 2008 09:01:52 +0100

cryptsetup (2:1.0.6~pre1-1) unstable; urgency=low

  [ Jonas Meurer ]
  * New upstream alpha release 1.0.6~pre1:

src/cryptsetup.c
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <inttypes.h>
#include <errno.h>
#include <unistd.h>
#include <assert.h>

#include <libcryptsetup.h>
#include <popt.h>

#include "../config.h"

#include "cryptsetup.h"

static int opt_verbose = 1;
static char *opt_cipher = NULL;
static char *opt_hash = DEFAULT_HASH;
static int opt_verify_passphrase = 0;
static char *opt_key_file = NULL;
static unsigned int opt_key_size = 0;
static int opt_key_slot = -1;
static uint64_t opt_size = 0;
static uint64_t opt_offset = 0;
static uint64_t opt_skip = 0;
static int opt_readonly = 0;
static int opt_iteration_time = 1000;
static int opt_batch_mode = 0;
static int opt_version_mode = 0;
static int opt_timeout = 0;
static int opt_tries = 3;
static int opt_align_payload = 0;

static const char **action_argv;
static int action_argc;

static int action_create(int arg);
static int action_remove(int arg);
static int action_resize(int arg);
static int action_status(int arg);
static int action_luksFormat(int arg);
static int action_luksOpen(int arg);
static int action_luksAddKey(int arg);
static int action_luksDelKey(int arg);
static int action_luksKillSlot(int arg);
static int action_luksRemoveKey(int arg);
static int action_isLuks(int arg);
static int action_luksUUID(int arg);
static int action_luksDump(int arg);

static struct action_type {
	const char *type;
 	int (*handler)(int);
	int arg;
	int required_action_argc;
	const char *arg_desc;
	const char *desc;
} action_types[] = {
	{ "create",	action_create, 0, 2, N_("<name> <device>"), N_("create device") },
	{ "remove",	action_remove, 0, 1, N_("<name>"), N_("remove device") },
	{ "reload",	action_create, 1, 2, N_("<name>"), N_("modify active device") },
	{ "resize",	action_resize, 0, 1, N_("<name>"), N_("resize active device") },
	{ "status",	action_status, 0, 1, N_("<name>"), N_("show device status") },
	{ "luksFormat",	action_luksFormat, 0, 1, N_("<device> [<new key file>]"), N_("formats a LUKS device") },
	{ "luksOpen",	action_luksOpen, 0, 2, N_("<device> <name> "), N_("open LUKS device as mapping <name>") },
	{ "luksAddKey",	action_luksAddKey, 0, 1, N_("<device> [<new key file>]"), N_("add key to LUKS device") },
	{ "luksRemoveKey", action_luksRemoveKey, 0, 1, N_("<device> [<key file]]"), N_("removes supplied key or key file from LUKS device") },
	{ "luksKillSlot",  action_luksKillSlot, 0, 2, N_("<device> <key slot>"), N_("wipes key with number <key slot> from LUKS device") },
	{ "luksDelKey",  action_luksDelKey, 0, 2, N_("<device> <key slot>"), N_("identical to luksKillSlot, but deprecated action name") },
	{ "luksUUID",	action_luksUUID, 0, 1, N_("<device>"), N_("print UUID of LUKS device") },
	{ "isLuks",	action_isLuks, 0, 1, N_("<device>"), N_("tests <device> for LUKS partition header") },
	{ "luksClose",	action_remove, 0, 1, N_("<name>"), N_("remove LUKS mapping") },
	{ "luksDump",	action_luksDump, 0, 1, N_("<device>"), N_("dump LUKS partition information") },
	{ NULL, NULL, 0, 0, NULL }
};

/* Interface Callbacks */
static int yesDialog(char *msg)
{
	int r = 0;
	if(isatty(0) && !opt_batch_mode) {
		char *answer=NULL;
	        size_t size=0;
		fprintf(stderr,"\nWARNING!\n========\n");
		fprintf(stderr,"%s\n\nAre you sure? (Type uppercase yes): ",msg);
		if(getline(&answer,&size,stdin) == -1)
			return 0;
		if(strcmp(answer,"YES\n") == 0)
			r = 1;
		free(answer);
	} else
		r = 1;
	return r;
}

static void cmdLineLog(int class, char *msg) {
    switch(class) {

    case CRYPT_LOG_NORMAL:
            fputs(msg, stdout);
            break;
    case CRYPT_LOG_ERROR:
            fputs(msg, stderr);
            break;
    default:
            fprintf(stderr, "Internal error on logging class for msg: %s", msg);
            break;
    }
}

static struct interface_callbacks cmd_icb = {
        .yesDialog = yesDialog,
        .log = cmdLineLog,
};

/* End ICBs */

static void show_status(int errcode)
{
	char error[256];

	if(!errcode) {
                fprintf(stderr, _("Command successful.\n"));
                return;
	}

	crypt_get_error(error, sizeof(error));
	if (!opt_verbose) {
		char *error_ = strerror_r(errcode, error, sizeof(error));
		if (error_ != error) {
			strncpy(error, error_, sizeof(error));
			error[sizeof error - 1] = '\0';
		}
	}

	fprintf(stderr, _("Command failed"));
	if (*error)
		fprintf(stderr, ": %s\n", error);
	else
		fputs(".\n", stderr);
	return;
}

static int action_create(int reload)
{
	struct crypt_options options = {
		.name = action_argv[0],
		.device = action_argv[1],
		.cipher = opt_cipher?opt_cipher:DEFAULT_CIPHER,
		.hash = opt_hash,
		.key_file = opt_key_file,
		.key_size = ((opt_key_size)?opt_key_size:DEFAULT_KEY_SIZE)/8,
		.key_slot = opt_key_slot,
		.passphrase_fd = 0,	/* stdin */
		.flags = 0,
		.size = opt_size,
		.offset = opt_offset,
		.skip = opt_skip,
		.timeout = opt_timeout,
		.tries = opt_tries,
		.icb = &cmd_icb,
	};
	int r;

	if (options.hash && strcmp(options.hash, "plain") == 0)
		options.hash = NULL;
	if (opt_verify_passphrase)
		options.flags |= CRYPT_FLAG_VERIFY;
	if (opt_readonly)
		options.flags |= CRYPT_FLAG_READONLY;

	if (reload)
		r = crypt_update_device(&options);
	else
		r = crypt_create_device(&options);
	if (r < 0)
		show_status(-r);
	return r;
}

static int action_remove(int arg)
{
	struct crypt_options options = {
		.name = action_argv[0],
		.icb = &cmd_icb,
	};
	int r;

	r = crypt_remove_device(&options);
	if (r < 0)
		show_status(-r);
	return r;
}

static int action_resize(int arg)
{
	struct crypt_options options = {
		.name = action_argv[0],
		.size = opt_size,
		.icb = &cmd_icb,
	};
	int r;

	r = crypt_resize_device(&options);
	if (r < 0)
		show_status(-r);
	return r;
}

static int action_status(int arg)
{
	struct crypt_options options = {
		.name = action_argv[0],
		.icb = &cmd_icb,
	};
	int r;

	r = crypt_query_device(&options);
	
	if (r < 0) {
		/* error */
		show_status(-r);
	} else if (r == 0) {
		/* inactive */
		printf("%s/%s is inactive.\n", crypt_get_dir(), options.name);
		r = 1;
	} else {
		/* active */
		printf("%s/%s is active:\n", crypt_get_dir(), options.name);
		printf("  cipher:  %s\n", options.cipher);
		printf("  keysize: %d bits\n", options.key_size * 8);
		printf("  device:  %s\n", options.device);
		printf("  offset:  %" PRIu64 " sectors\n", options.offset);
		printf("  size:    %" PRIu64 " sectors\n", options.size);
		if (options.skip)
			printf("  skipped: %" PRIu64 " sectors\n", options.skip);
		printf("  mode:    %s\n", (options.flags & CRYPT_FLAG_READONLY)
		                           ? "readonly" : "read/write");
		crypt_put_options(&options);
		r = 0;
	}
	return r;
}

static int action_luksFormat(int arg)
{
	struct crypt_options options = {
		.key_size = (opt_key_size != 0 ? opt_key_size : DEFAULT_LUKS_KEY_SIZE) / 8,
		.key_slot = opt_key_slot,
		.device = action_argv[0],
		.cipher = opt_cipher?opt_cipher:DEFAULT_LUKS_CIPHER,
		.new_key_file = action_argc > 1 ? action_argv[1] : NULL,
		.flags = opt_verify_passphrase ? CRYPT_FLAG_VERIFY : (!opt_batch_mode?CRYPT_FLAG_VERIFY_IF_POSSIBLE :  0),
		.iteration_time = opt_iteration_time,
		.timeout = opt_timeout,
		.align_payload = opt_align_payload,
		.icb = &cmd_icb,
	};

	int r = 0; char *msg = NULL;

	if(asprintf(&msg, _("This will overwrite data on %s irrevocably."), options.device) == -1) {
		fputs(_("memory allocation error in action_luksFormat"), stderr);
	} else {
		r = yesDialog(msg) ? crypt_luksFormat(&options) : -EINVAL;
		free(msg);
		show_status(-r);
	}
	return r;
}

static int action_luksOpen(int arg)
{
	struct crypt_options options = {
		.name = action_argv[1],
		.device = action_argv[0],
		.key_file = opt_key_file,
		.timeout = opt_timeout,
		.tries = opt_tries,
		.icb = &cmd_icb,
	};
	int r; 

	opt_verbose = 1;
	options.flags = 0;
	if (opt_readonly)
		options.flags |= CRYPT_FLAG_READONLY;
	r = crypt_luksOpen(&options);
	show_status(-r);
	return r;
}

static int action_luksDelKey(int arg)
{
    fprintf(stderr,"luksDelKey is a deprecated action name.\nPlease use luksKillSlot that is completely identicaly (but more clear in naming)\n"); 
    return action_luksKillSlot(arg);
}

static int action_luksKillSlot(int arg)
{
	struct crypt_options options = {
		.device = action_argv[0],
		.key_slot = atoi(action_argv[1]),
		.key_file = opt_key_file,
		.timeout = opt_timeout,
		.flags = !opt_batch_mode?CRYPT_FLAG_VERIFY_ON_DELKEY : 0,
		.icb = &cmd_icb,
	};
	int r; 

	opt_verbose = 1;
	r = crypt_luksKillSlot(&options);
	show_status(-r);
	return r;
}

static int action_luksRemoveKey(int arg)
{
	struct crypt_options options = {
		.device = action_argv[0],
		.new_key_file = action_argc>1?action_argv[1]:NULL,
		.key_file = opt_key_file,
		.timeout = opt_timeout,
		.flags = !opt_batch_mode?CRYPT_FLAG_VERIFY_ON_DELKEY : 0,
		.icb = &cmd_icb,
	};
	int r; 

	opt_verbose = 1;
	r = crypt_luksRemoveKey(&options);
	show_status(-r);
	return r;
}

static int action_luksAddKey(int arg)
{
	struct crypt_options options = {
		.device = action_argv[0],
		.new_key_file = action_argc>1?action_argv[1]:NULL,
		.key_file = opt_key_file,
		.key_slot = opt_key_slot,
		.flags = opt_verify_passphrase ? CRYPT_FLAG_VERIFY : (!opt_batch_mode?CRYPT_FLAG_VERIFY_IF_POSSIBLE : 0),
		.iteration_time = opt_iteration_time,
		.timeout = opt_timeout,
		.icb = &cmd_icb,
	};
	int r; 

	opt_verbose = 1;
	r = crypt_luksAddKey(&options);
	show_status(-r);
	return r;
}

static int action_isLuks(int arg)
{
	struct crypt_options options = {
		.device = action_argv[0],
		.icb = &cmd_icb,
	};
	return crypt_isLuks(&options);
}

static int action_luksUUID(int arg)
{
	struct crypt_options options = {
		.device = action_argv[0],
		.icb = &cmd_icb,
	};
	int r;

	r = crypt_luksUUID(&options);
	if (r < 0)
		show_status(-r);
	return r;
}

static int action_luksDump(int arg)
{
	struct crypt_options options = {
		.device = action_argv[0],
		.icb = &cmd_icb,
	};
	int r; 

	r = crypt_luksDump(&options);
	if (r < 0)
		show_status(-r);
	return r;
}

static void usage(poptContext popt_context, int exitcode,
                  const char *error, const char *more)
{
	poptPrintUsage(popt_context, stderr, 0);
	if (error)
		fprintf(stderr, "%s: %s\n", more, error);
	exit(exitcode);
}

static void help(poptContext popt_context, enum poptCallbackReason reason,
                 struct poptOption *key, const char * arg, void *data)
{
	if (key->shortName == '?') {
		struct action_type *action;

		fprintf(stdout, "%s\n",PACKAGE_STRING);

		poptPrintHelp(popt_context, stdout, 0);

		printf(_("\n"
			 "<action> is one of:\n"));

		for(action = action_types; action->type; action++)
			printf("\t%s %s - %s\n", action->type, _(action->arg_desc), _(action->desc));
		
		printf(_("\n"
			 "<name> is the device to create under %s\n"
			 "<device> is the encrypted device\n"
			 "<key slot> is the LUKS key slot number to modify\n"
			 "<key file> optional key file for the new key for luksAddKey action\n"),
			crypt_get_dir());
		exit(0);
	} else
		usage(popt_context, 0, NULL, NULL);
}                 

int main(int argc, char **argv)
{
	static char *popt_tmp;
	static struct poptOption popt_help_options[] = {
		{ NULL,    '\0', POPT_ARG_CALLBACK, help, 0, NULL,                         NULL },
		{ "help",  '?',  POPT_ARG_NONE,     NULL, 0, N_("Show this help message"), NULL },
		{ "usage", '\0', POPT_ARG_NONE,     NULL, 0, N_("Display brief usage"),    NULL },
		POPT_TABLEEND
	};
	static struct poptOption popt_options[] = {
		{ NULL,                '\0', POPT_ARG_INCLUDE_TABLE,                      popt_help_options,      0, N_("Help options:"),                                                   NULL },
		{ "verbose",           'v',  POPT_ARG_NONE,                               &opt_verbose,           0, N_("Shows more detailed error messages"),                              NULL },
		{ "cipher",            'c',  POPT_ARG_STRING | POPT_ARGFLAG_SHOW_DEFAULT, &opt_cipher,            0, N_("The cipher used to encrypt the disk (see /proc/crypto)"),          NULL },
		{ "hash",              'h',  POPT_ARG_STRING | POPT_ARGFLAG_SHOW_DEFAULT, &opt_hash,              0, N_("The hash used to create the encryption key from the passphrase"),  NULL },
		{ "verify-passphrase", 'y',  POPT_ARG_NONE,                               &opt_verify_passphrase, 0, N_("Verifies the passphrase by asking for it twice"),                  NULL },
		{ "key-file",          'd',  POPT_ARG_STRING,                             &opt_key_file,          0, N_("Read the key from a file (can be /dev/random)"),                   NULL },
		{ "key-size",          's',  POPT_ARG_INT    | POPT_ARGFLAG_SHOW_DEFAULT, &opt_key_size,          0, N_("The size of the encryption key"),                                  N_("BITS") },
		{ "key-slot",          'S',  POPT_ARG_INT,                                &opt_key_slot,          0, N_("Slot number for new key (default is first free)"),      NULL },
		{ "size",              'b',  POPT_ARG_STRING,                             &popt_tmp,              1, N_("The size of the device"),                                          N_("SECTORS") },
		{ "offset",            'o',  POPT_ARG_STRING,                             &popt_tmp,              2, N_("The start offset in the backend device"),                          N_("SECTORS") },
		{ "skip",              'p',  POPT_ARG_STRING,                             &popt_tmp,              3, N_("How many sectors of the encrypted data to skip at the beginning"), N_("SECTORS") },
		{ "readonly",          'r',  POPT_ARG_NONE,                               &opt_readonly,          0, N_("Create a readonly mapping"),                                       NULL },
		{ "iter-time",         'i',  POPT_ARG_INT,                                &opt_iteration_time,    0, N_("PBKDF2 iteration time for LUKS (in ms)"),
		  N_("msecs") },
		{ "batch-mode",        'q',  POPT_ARG_NONE,                               &opt_batch_mode,        0, N_("Do not ask for confirmation"),                                     NULL },
		{ "version",        '\0',  POPT_ARG_NONE,                               &opt_version_mode,        0, N_("Print package version"),                                     NULL },
 		{ "timeout",           't',  POPT_ARG_INT,                                &opt_timeout,           0, N_("Timeout for interactive passphrase prompt (in seconds)"),          N_("secs") },
  		{ "tries",             'T',  POPT_ARG_INT,                                &opt_tries,             0, N_("How often the input of the passphrase can be retried"),            NULL },
 		{ "align-payload",     '\0',  POPT_ARG_INT,                                &opt_align_payload,     0, N_("Align payload at <n> sector boundaries - for luksFormat"),         N_("SECTORS") },
		POPT_TABLEEND
	};
	poptContext popt_context;
	struct action_type *action;
	char *aname;
	int r;
	const char *null_action_argv[] = {NULL};

	setlocale(LC_ALL, "");
	bindtextdomain(GETTEXT_PACKAGE, LOCALEDIR);
	textdomain(GETTEXT_PACKAGE);

	popt_context = poptGetContext(PACKAGE, argc, (const char **)argv,
	                              popt_options, 0);
	poptSetOtherOptionHelp(popt_context,
	                       N_("[OPTION...] <action> <action-specific>]"));

	while((r = poptGetNextOpt(popt_context)) > 0) {
		unsigned long long ull_value;
		char *endp;

		ull_value = strtoull(popt_tmp, &endp, 0);
		if (*endp || !*popt_tmp)
			r = POPT_ERROR_BADNUMBER;

		switch(r) {
			case 1:
				opt_size = ull_value;
				break;
			case 2:
				opt_offset = ull_value;
				break;
			case 3:
				opt_skip = ull_value;
				break;
		}

		if (r < 0)
			break;
	}

	if (r < -1)
		usage(popt_context, 1, poptStrerror(r),
		      poptBadOption(popt_context, POPT_BADOPTION_NOALIAS));
	if (opt_version_mode) {
	        printf("%s %s\n", PACKAGE_NAME, PACKAGE_VERSION);
		exit(0);
	}
	 
	if (opt_key_size % 8)
		usage(popt_context, 1,
		      _("Key size must be a multiple of 8 bits"),
		      poptGetInvocationName(popt_context));
	
	if (!(aname = (char *)poptGetArg(popt_context)))
		usage(popt_context, 1, _("Argument <action> missing."),
		      poptGetInvocationName(popt_context));
	for(action = action_types; action->type; action++)
		if (strcmp(action->type, aname) == 0)
			break;
	if (!action->type)
		usage(popt_context, 1, _("Unknown action."),
		      poptGetInvocationName(popt_context));

	action_argc = 0;
	action_argv = poptGetArgs(popt_context);
	/* Make return values of poptGetArgs more consistent in case of remaining argc = 0 */
	if(!action_argv) 
		action_argv = null_action_argv;
	
	/* Count args, somewhat unnice, change? */
	while(action_argv[action_argc] != NULL)
		action_argc++;

	if(action_argc < action->required_action_argc) {
		char buf[128];
		snprintf(buf, 128,_("%s: requires %s as arguments"), action->type, action->arg_desc);
		usage(popt_context, 1, buf,
		      poptGetInvocationName(popt_context));
	}	
	return action->handler(action->arg);
}

// Local Variables:
// c-basic-offset: 8
// indent-tabs-mode: nil
// End:
